### ISSUE TITLE
ccache reports missing temporary files

### ISSUE THREAD

**Issue by fgouget** [ccache/ccache#480]
I have been using ccache for a long time without trouble. However I now have a build VM and in it the builds sometimes fail with an error like the one below:

ccache: error: Failed to open /home/winetest/.ccache/tmp/tmp.cpp_stderr.fgtbdebian10.2470.JcWDlw: No such file or directory

Of course these builds are parallelized so there is the potential for race conditions in ccache.

The filesystem is ext4:
/dev/sda1 on / type ext4 (rw,relatime,errors=remount-ro)

$ cat .ccache/ccache.conf 
max_size = 5G

ccache is Debian 10's package, version 3.6-1. Depending on the build the compiler is one of gcc (4:8.3.0-1), i686-w64-mingw32-gcc (8.3.0-6+21.3\~deb10u1) or x86_64-w64-mingw32-gcc (8.3.0-6+21.3\~deb10u1).

**Comment by jrosdahl** [comment]
This sounds like #154 which I was never able to reproduce. Just running many compilations in parallel is not enough since many people do that all the time without problem (the temporary file is created using `mkstemp` so libc guarantees that it's unique).

The code in question basically works like this:

1. Create a unique temporary file.
2. Call the real compiler while redirecting standard error output to the file.
3. Hash the temporary file.
4. Delete the temporary file.

If the file is missing in step 3, ccache fails fatally.

I've been thinking some more about this now. The problem has to be that somebody removed the temporary file during step 2. ccache itself scans `$ccache_dir/tmp` once an hour (slightly simplified) and removes temporary files that are older than one hour. I see two things that could trigger the error:

1. Perform a compilation that takes longer than hour (perhaps pause it) and compile something else after an hour. This will trigger the internal cleanup mechanism and then ccache will fail when the long compilation finishes.
2. Change the system clock forward (more than one hour or so) during a build so that a new ccache invocation will remove the temporary file created by an ongoing ccache invocation.

In either case I think that it would make sense not to fail fatally if the temporary file is missing, only fall back to running the real compiler.

**Comment by fgouget** [comment]
> Change the system clock forward (more than one hour or so) during a build so that a new
> ccache invocation will remove the temporary file created by an ongoing ccache invocation.

Now that is interesting. This system uses live snapshots of the build VM which means the guest system clock is out of date. It goes like this:
  1. Revert the VM to the live snapshot
  2. In the guest update the source
  3. Rebuild
  4. Recreate the live snapshot

This happens daily which means the guest system clock is typically running behind by around 24 hours. So if I understand you correctly and the clock does get moved forward during the build, this may cause a new ccache process to delete the temporary files created by another ccache instance that started just before the system clock update.

I'm adding a 1b step that updates the clock before the build starts and hopefully that will stop these random build failures.

Note also that in many countries the clock is moved forward by one hour once a year during the daylight saving time change. However ccache presumably uses the UTC time which should not be affected.

Still I think it would make sense for ccache to be more robust with regards to system time changes.

**Comment by Romain-Geissler-1A** [comment]
Hi @jrosdahl,

Sorry to jump back into a closed issue, but I think there is still some room for improvement.

It happens that recently in my company we started to hit this issue more and more. We think it is linked to the fact that more and more of our builds are now targetting a single common ccache folder, which is mounted through NFS between hundreads and hundreads of machines. As it's written many times in the code, on NFS file operations aren't as atomic as we could hope for a local filesystem, and most likely problems are coming down from here. Note that we are using more and more docker containers for that, with containers having fixed hostname rather than generated "random" ones, which decreases entropy of the names you generate.

In the end what we did is that we saw your commit, we upgrade ccache, and now are build stop failing "randomly" (still it happened very often). That's cool, but it is just a workaround, in the end the goal of ccache is to avoid rebuilding as much as possible, and here we have cases where we know we rebuild just because we use NFS where atomicity of operations is not guaranteed, not because there is an actual code change. And the more machines we have, the more collision there are, and the more useless rebuilds there are.

So I propose we limit the amount of collision for the NFS case by adding more entropy in the temporary files names. In legacy_util.cpp, function tmp_string you use both hostname and pid to add entropy. In the context of recent linux with container support, using just the hostname and the pid is not enough to try to be "unique". For the hostname I would prepend it with the Linux UTS namespace id. For the PID, I would prepend it with the Linux PID namespace id. How to access namespace ids from C/C++ ? Just call "stat("/proc/self/ns/uts")" and "stat(/proc/self/ns/pid")" if these files do exists, and there the namespace id is actually the inode number "st_ino". Old Linux will not have these files, so you shall cope with it. Make sure you use "stat" and not "lstat", these are symlinks and we want the inode of the file pointed too. Doing that would reduce collisions between ccache run from containers using the same hostname (as we do).

It happens that I also had these issues sometimes one the same machine using the same container, so namespace were the same. I guess this may come from big project spawning a huge amount of very small processes very quickly, and somehow reaching pid_max quickly and reverting back to small pids. So what I would do as well is maybe add more entropy by not only using the pid of the current process, but also the parent pid getppid().

What do you think about the collision reduction mechanisms ?

Cheers,
Romain

**Comment by jrosdahl** [comment]
Hi,

> Sorry to jump back into a closed issue, but I think there is still some room for improvement.

Since this issue was about temporary files being removed when updating the system clock, why not create a new separate issue for your improvement suggestion? I don't think that it's a good idea to discuss a separate issue here.

**Comment by jrosdahl** [comment]
>  why not create a new separate issue for your improvement suggestion?

@Romain-Geissler-1A: I've opened a separate issue for discussing this topic now, so please continue there: #512.
